// @nervos/appchain-truffle-migrate/dist/contract/contract.self.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ethJSABI = require('ethjs-abi');
var BlockchainUtils = require('truffle-blockchain-utils');
var Web3 = require('@nervos/web3');
var StatusError = require('./statuserror.js');
var log = console.log.bind(console, '-- contract.js --\n');
var Provider = (function () {
    function Provider(provider) {
        this.provider = provider;
    }
    Provider.prototype.send = function () {
        return this.provider.send.apply(this.provider, arguments);
    };
    Provider.prototype.sendAsync = function () {
        return this.provider.sendAsync.apply(this.provider, arguments);
    };
    return Provider;
}());
var Contract = (function () {
    function Contract(contract) {
        if (typeof contract == 'string') {
            var constructor = this.constructor;
            var address = contract;
            var contract_class = constructor.web3.eth.contract(this.abi);
            contract = contract_class.at(address);
        }
        this.contract = contract;
    }
    Contract.clone = function (binary) { };
    return Contract;
}());
exports.default = Contract;
