// @nervos/appchain-truffle-migrate/dist/deploy.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var _this = this;
Object.defineProperty(exports, "__esModule", { value: true });
var log_1 = require("./log");
var contract_utils_1 = require("./contract_utils");
var assertAbiIsStored = function (receipt, web3, address) { return __awaiter(_this, void 0, void 0, function () {
    var err, abi;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                err = receipt.errorMessage;
                if (!(err === null)) return [3, 2];
                return [4, web3.appchain.getAbi(address)];
            case 1:
                abi = _a.sent();
                if (abi === '0x') {
                    err = 'store abi failure';
                }
                else {
                    console.log('store abi success');
                    return [2];
                }
                _a.label = 2;
            case 2: throw err;
        }
    });
}); };
var pollingTransationReceipt = function (address, web3, hash) { return __awaiter(_this, void 0, void 0, function () {
    var i, receipt, err_1;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                i = 0;
                receipt = null;
                _a.label = 1;
            case 1:
                if (!(i < 10)) return [3, 9];
                return [4, web3.appchain.getTransactionReceipt(hash)];
            case 2:
                receipt = _a.sent();
                if (!receipt) return [3, 7];
                log_1.default('轮询结果', receipt);
                _a.label = 3;
            case 3:
                _a.trys.push([3, 5, , 6]);
                return [4, assertAbiIsStored(receipt, web3, address)];
            case 4:
                _a.sent();
                return [3, 6];
            case 5:
                err_1 = _a.sent();
                console.error(err_1);
                return [3, 6];
            case 6: return [2];
            case 7: return [4, new Promise(function (resolve, reject) {
                    setTimeout(function () {
                        resolve();
                    }, 1000);
                })];
            case 8:
                _a.sent();
                i += 1;
                log_1.default('获取交易收据次数', i);
                return [3, 1];
            case 9:
                console.error('fetch transaction receipt overtime');
                return [2];
        }
    });
}); };
var storeAbiToBlockchain = function (contractInfo, web3, address) { return __awaiter(_this, void 0, void 0, function () {
    var abi, validUntilBlock, chainId, nonce, value, version, quota, privateKey, abibytes, data, tx, res, hash, err_2;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                abi = contractInfo.abi, validUntilBlock = contractInfo.validUntilBlock, chainId = contractInfo.chainId, nonce = contractInfo.nonce, value = contractInfo.value, version = contractInfo.version, quota = contractInfo.quota, privateKey = contractInfo.privateKey;
                abibytes = contract_utils_1.fromUtf8(JSON.stringify(abi));
                data = address + abibytes;
                tx = {
                    from: 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',
                    to: 'ffffffffffffffffffffffffffffffffff010001',
                    quota: quota,
                    version: version,
                    value: value,
                    nonce: nonce,
                    data: data,
                    validUntilBlock: validUntilBlock,
                    chainId: chainId,
                    privateKey: privateKey,
                };
                _a.label = 1;
            case 1:
                _a.trys.push([1, 3, , 4]);
                return [4, web3.appchain.sendTransaction(tx).catch(console.error)];
            case 2:
                res = _a.sent();
                hash = res.hash;
                res = pollingTransationReceipt(address, web3, hash);
                return [3, 4];
            case 3:
                err_2 = _a.sent();
                console.error(err_2);
                return [3, 4];
            case 4: return [2];
        }
    });
}); };
var deployContract = function (contractInfo, web3, artifact, args) { return __awaiter(_this, void 0, void 0, function () {
    var bytecode, privateKey, from, nonce, quota, value, chainId, version, params, res, errDeploy, address;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                log_1.default('deployContract');
                bytecode = contractInfo.bytecode, privateKey = contractInfo.privateKey, from = contractInfo.from, nonce = contractInfo.nonce, quota = contractInfo.quota, value = contractInfo.value, chainId = contractInfo.chainId, version = contractInfo.version;
                params = { privateKey: privateKey, from: from, nonce: nonce, quota: quota, value: value, chainId: chainId, version: version };
                return [4, web3.appchain.deploy(bytecode, params)];
            case 1:
                res = _a.sent();
                errDeploy = res.errorMessage;
                if (errDeploy) {
                    throw errDeploy;
                }
                address = res.contractAddress;
                console.log('contract deployed successful, address:', address);
                return [4, storeAbiToBlockchain(contractInfo, web3, address)];
            case 2:
                _a.sent();
                artifact.address = address;
                return [2];
        }
    });
}); };
var deploy = function (contractInfo, web3, artifact, args) { return __awaiter(_this, void 0, void 0, function () {
    var validUntilBlock, res, num;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                validUntilBlock = contractInfo.validUntilBlock;
                if (!(validUntilBlock === undefined)) return [3, 2];
                return [4, web3.appchain.getBlockNumber()];
            case 1:
                res = _a.sent();
                num = Number(res);
                contractInfo.validUntilBlock = num + 88;
                _a.label = 2;
            case 2: return [4, deployContract(contractInfo, web3, artifact, args)];
            case 3:
                _a.sent();
                return [2, artifact];
        }
    });
}); };
exports.default = deploy;
